perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -4741,8 +4753,39 @@ public void removeGroupUnion(PerunSession sess, Group resultGroup, Group operand
 
 		// recalculates statuses of members in result group
 		for (Member member : getGroupMembers(sess, resultGroup)) {
-			recalculateMemberGroupStatusRecursively(sess, member, resultGroup);
+			recalculateMemberGroupStatusRecursively(sess, member, resultGroup, previousStatuses);
+		}
+	}
+
+	/**
+	 * Returns previous total member-group statuses of members in all result groups of the given group.
+	 *
+	 * @param sess perun session
+	 * @param group group
+	 * @param members members
+	 * @param previousStatuses previous statuses, empty in at the beginning
+	 * @return previous statuses of members in all result groups of the given group
+	 */
+	private Map<Pair<Group, Member>, MemberGroupStatus> getPreviousStatuses(PerunSession sess, Group group, List<Member> members, Map<Pair<Group, Member>, MemberGroupStatus> previousStatuses) {
+
+		for (Member member : members) {
+			Pair<Group, Member> groupMemberPair = new Pair<>(group, member);
+			previousStatuses.put(groupMemberPair, getTotalMemberGroupStatus(sess, member, group));
```

Comment:
I think this could be a performance problem. If we consider that some groups have thousands of users, this would take a lot of time. I think we could use here some method that would retrieve statuses of all members of this group (create method `getTotalGroupStatusForMembers`.

Reply:
Good idea, added, thanks.